lib/logutils/firstandinterval.go

```diff
@@ -0,0 +1,296 @@
+// Copyright (c) 2021 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package logutils
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	fieldLogSkipped = "logs-skipped"
+	fieldLogNextLog = "next-log"
+)
+
+// NewFirstAndIntervalLogger returns a FirstAndIntervalLogger which can be used for interval logging.
+//
+// Methods are essentially the same as the logrus logging methods, but there is no Panic or Fatal log since these don't
+// make much sense for interval logging.
+//
+// The first log is always processed.  Subsequent logs are only processed if at least the interval time has passed since
+// the last log request. Processed logs contain information about the number of skipped logs and the expected earliest
+// time of the next log. Force() may be used to force the next log to be processed.  Note that sufficient log level is
+// still required for a processed log to actually be written.
+//
+// Typical use might be as follows:
+//
+//   logger := NewFirstAndIntervalLogger(5 * time.Minute).WithField("key": "my-key")
+//   for {
+//     logger.Infof("Checking some stuff: %s", myStuff)
+//     complete = doSomeStuff()
+//     if complete {
+//       break
+//     }
+//   }
+//
+//   // Use force to ensure our final log is printed and it contains the summary info about the number of skipped logs.
+//   logger.Force().Info("Finished checking stuff")
+//
+func NewFirstAndIntervalLogger(interval time.Duration, logger *logrus.Logger) *FirstAndIntervalLogger {
+	if logger == nil {
+		logger = logrus.StandardLogger()
+	}
+	return &FirstAndIntervalLogger{
+		nextLog: time.Now(),
+		interval: interval,
+		entry:    logrus.NewEntry(logger),
+	}
+}
+
+type FirstAndIntervalLogger struct {
+	nextLog  time.Time
+
+	// Interval for logging.
+	interval time.Duration
+
+	// The logrus entry used for writing the log.
+	entry    *logrus.Entry
+
+	// The number skipped since the last processed log.
+	skipped  int
+
+	// Whether to force the next log to be processed.
+	force    bool
+
+	// Lock used to access to this data. This lock is never held while writing a log.
+	lock     sync.RWMutex
```

Comment:
`sync.Mutex` is a bit faster if you're not using the `RLock()` function

Reply:
Changed.